app/app_main.c

```diff
@@ -2694,100 +2710,56 @@ static ACVP_RESULT app_kdf135_ssh_handler(ACVP_TEST_CASE *test_case)
-    long msglen = -1;
-    BIGNUM *bn_e = NULL;    
+    unsigned int bitlen1, bitlen2, bitlen3, bitlen4, keylen;
     
     if (!test_case) {
-        return ACVP_INVALID_ARG;
+        rv = ACVP_INVALID_ARG;
+        goto err;
+    }
+    tc = test_case->tc.rsa_keygen;
+    
+    rsa = FIPS_rsa_new();
+    bitlen1 = tc->bitlen1;
+    bitlen2 = tc->bitlen2;
+    bitlen3 = tc->bitlen3;
+    bitlen4 = tc->bitlen4;
+    keylen = tc->modulo;
+    
+    int rc = rsa_generate_key_internal(&rsa->p, &rsa->q, &rsa->n, &rsa->d,
+                                       tc->seed, tc->seed_len,
+                                       bitlen1, bitlen2, bitlen3, bitlen4,
+                                       tc->e, keylen, NULL);
+    if (rc != 1) {
+        printf("Error generating key\n");
+        rv = ACVP_CRYPTO_MODULE_FAIL;
+        goto err;
     }
+    
+    tc->p = BN_dup(rsa->p);
+    tc->q = BN_dup(rsa->q);
+    tc->n = BN_dup(rsa->n);
+    tc->d = BN_dup(rsa->d);
 
-    tc = test_case->tc.rsa;
-    switch(tc->mode) {
-    case ACVP_RSA_MODE_KEYGEN:
-        switch (tc->rand_pq) {
-        case RSA_RAND_PQ_B34: // "provPC"
-            rsa = RSA_new();
-            if(tc->info_gen_by_server) {
-                exponent = tc->keygen_tc->e;
-                bitlen1 = tc->keygen_tc->bitlen1;
-                bitlen2 = tc->keygen_tc->bitlen2;
-                bitlen3 = tc->keygen_tc->bitlen3;
-                bitlen4 = tc->keygen_tc->bitlen4;
-                seed_len = tc->keygen_tc->seed_len;
-                seed = tc->keygen_tc->seed;
-            } else {
-                exponent = BN_new();
-                m = RSA_F4;
-                seed_len = 28;
-                seed = calloc(seed_len, sizeof(char));
-                if (!BN_set_word(exponent, m)) {
-                    printf("Bignum API fail\n");
-                    return ACVP_CRYPTO_MODULE_FAIL;
-                }
-                if (RAND_bytes(seed, seed_len) <= 0) {
-                    printf("RAND API fail\n");
-                    return ACVP_CRYPTO_MODULE_FAIL;
-                }
-                bitlen1 = 160;
-                bitlen2 = 272;
-                bitlen3 = 240;
-                bitlen4 = 184;
-                keylen = 2048;
-            }
-            break;
-        case RSA_RAND_PQ_B32: // "provRP"
-        case RSA_RAND_PQ_B33: // "probRP"
-        case RSA_RAND_PQ_B35: // "bothPC"
-        case RSA_RAND_PQ_B36: // "probPC"
-        default:
-            break;
-        }
-        if(rsa_generate_key_internal(&rsa->p, &rsa->q, &rsa->n, &rsa->d,
-                                      seed, seed_len,
-                                      bitlen1, bitlen2, bitlen3, bitlen4,
-                                      exponent, keylen, NULL) != 1) {
-            return ACVP_CRYPTO_MODULE_FAIL;
-        }
-        tc->keygen_tc->p = rsa->p;
-        tc->keygen_tc->q = rsa->q;
-        tc->keygen_tc->n = rsa->n;
-        tc->keygen_tc->d = rsa->d;
+    RSA_free(rsa);
```

Comment:
Probably need to be consistent and use the FOM calls for free since we use FOM call FIPS_rsa_new()